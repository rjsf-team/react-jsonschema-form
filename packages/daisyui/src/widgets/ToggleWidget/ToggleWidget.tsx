import { ChangeEvent, useCallback } from 'react';
import { FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';

/** The `ToggleWidget` component renders a toggle switch input with DaisyUI styling
 *
 * Features:
 * - Provides a visual toggle switch rather than a standard checkbox
 * - Supports different sizes through options (sm, md, lg)
 * - Handles required, disabled, and readonly states
 * - Manages focus and blur events for accessibility
 * - Includes an optional label from options
 *
 * @param props - The `WidgetProps` for this component
 */
export default function ToggleWidget<
  T = any,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = any,
>({ id, value, required, disabled, readonly, autofocus, onChange, onFocus, onBlur, options }: WidgetProps<T, S, F>) {
  /** Handle change events from the toggle input
   *
   * @param event - The change event
   */
  const _onChange = useCallback(
    ({ target: { checked } }: ChangeEvent<HTMLInputElement>) => onChange(checked),
    [onChange],
  );

  /** Handle focus events
   */
  const handleFocus = useCallback(() => {
    if (onFocus) {
      onFocus(id, value);
    }
  }, [onFocus, id, value]);

  /** Handle blur events
   */
  const handleBlur = useCallback(() => {
    if (onBlur) {
      onBlur(id, value);
    }
  }, [onBlur, id, value]);

  // Get size from options or use default "md"
  const { size = 'md' } = options;

  // Only add size class if it's not the default size
  const sizeClass = size !== 'md' ? `toggle-${size}` : '';

  return (
    <div className='form-control'>
      <label className='cursor-pointer label my-auto'>
        <input
          type='checkbox'
          id={id}
          checked={value}
          required={required}
          disabled={disabled || readonly}
          autoFocus={autofocus}
          onChange={_onChange}
          onFocus={handleFocus}
          onBlur={handleBlur}
          className={`toggle ${sizeClass}`}
        />
        <span className='label-text'>{options.label}</span>
      </label>
    </div>
  );
}
