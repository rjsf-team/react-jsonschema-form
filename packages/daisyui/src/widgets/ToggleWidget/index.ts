export { default } from './ToggleWidget';
export * from './ToggleWidget';
