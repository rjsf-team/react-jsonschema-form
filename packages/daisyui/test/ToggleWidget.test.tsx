import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';

import ToggleWidget from '../src/widgets/ToggleWidget/ToggleWidget';
import { makeWidgetMockProps } from './helpers/createMocks';

describe('ToggleWidget', () => {
  test('renders correctly', () => {
    const tree = renderer.create(<ToggleWidget {...makeWidgetMockProps()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders with custom label', () => {
    const tree = renderer
      .create(
        <ToggleWidget
          {...makeWidgetMockProps({
            label: 'Custom Toggle Label',
          })}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders disabled state correctly', () => {
    const tree = renderer
      .create(
        <ToggleWidget
          {...makeWidgetMockProps({
            disabled: true,
          })}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('calls onChange when toggled', () => {
    const onChange = jest.fn();
    const { getByRole } = render(
      <ToggleWidget
        {...makeWidgetMockProps({
          onChange,
          value: false,
        })}
      />
    );

    const toggle = getByRole('checkbox');
    fireEvent.click(toggle);

    expect(onChange).toHaveBeenCalledWith(true);
  });

  test('renders with correct checked state', () => {
    const { getByRole } = render(
      <ToggleWidget
        {...makeWidgetMockProps({
          value: true,
        })}
      />
    );

    const toggle = getByRole('checkbox');
    expect(toggle).toHaveProperty('checked', true);
  });
});
