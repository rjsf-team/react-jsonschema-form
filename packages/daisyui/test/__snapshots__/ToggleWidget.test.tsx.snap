// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleWidget renders correctly 1`] = `
<div
  className="form-control"
>
  <label
    className="cursor-pointer label my-auto"
  >
    <input
      autoFocus={true}
      checked="test-value"
      className="toggle "
      disabled={false}
      id="test-id"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="checkbox"
    />
    <span
      className="label-text"
    />
  </label>
</div>
`;

exports[`ToggleWidget renders disabled state correctly 1`] = `
<div
  className="form-control"
>
  <label
    className="cursor-pointer label my-auto"
  >
    <input
      autoFocus={true}
      checked="test-value"
      className="toggle "
      disabled={true}
      id="test-id"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="checkbox"
    />
    <span
      className="label-text"
    />
  </label>
</div>
`;

exports[`ToggleWidget renders with custom label 1`] = `
<div
  className="form-control"
>
  <label
    className="cursor-pointer label my-auto"
  >
    <input
      autoFocus={true}
      checked="test-value"
      className="toggle "
      disabled={false}
      id="test-id"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="checkbox"
    />
    <span
      className="label-text"
    />
  </label>
</div>
`;
